### DISCUSSION TITLE
TypeVarTuple with ParamSpec

### DISCUSSION THREAD

**Discussion by cdce8p** [python/typing#1194] [Q&A]
I'm trying to find a better solution for `call_job` (below). Currently, `*args` is typed as `Any`.

```py
from typing import Callable, Generic, Any
from typing_extensions import ParamSpec

P = ParamSpec("P")

class Job(Generic[P]):
    def __init__(self, target: Callable[P, None]) -> None:
        self.target = target

def call_job(job: Job[...], *args: Any) -> None:
    job.target(*args)
```

**Idea**
With PEP 646 I'm wondering if it's possible to pass an unpacked TypeVarTuple as argument to `Job`, similar to how it works with Callables. Basically, `[*Ts]` is just treated as a sequence of types. https://peps.python.org/pep-0646/#type-variable-tuples-with-callable

```py
from typing_extensions import TypeVarTuple

Ts = TypeVarTuple("Ts")

def call_job(job: Job[[*Ts]], *args: *Ts) -> None:
    reveal_type(job.target)
    job.target(*args)
```

**Problem**
Is this supposed to work? In a discussion in the pyright repo, Eric mentioned that this is unsupported: https://github.com/microsoft/pyright/discussions/3310#discussioncomment-2526448. Pyright does emit an error, although interestingly `job.target` is still inferred correctly.
```
  test.py:18:24 - error: TypeVarTuple not allowed in this context
  test.py:19:17 - information: Type of "job.target" is "(*Ts@call_job) -> None"
```

**Comment by JelleZijlstra** [comment]
If your `Job` class is made generic over a TypeVarTuple inside of a ParamSpec, I believe this should work. However, you can't use a TypeVarTuple as the value for a ParamSpec.

**Reply by mschurr** [reply]
Maybe I'm missing something, but why not:
```
def call_job(job: Job[P], *args: P.args, **kwargs: P.kwargs) -> None:
    job.target(*args, **kwargs)
```

**Reply by cdce8p** [reply]
> Maybe I'm missing something, but why not:
> 
> ```
> def call_job(job: Job[P], *args: P.args, **kwargs: P.kwargs) -> None:
>     job.target(*args, **kwargs)
> ```

There are a number of functions, even in the stdlib, that don't accept `**kwargs`. Thus the normal ParamSpec usage doesn't work.
E.g. `asyncio.call_soon`, `asyncio.call_later`. https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_soon
